node/src/exchange/server.rs

```diff
@@ -1,52 +1,396 @@
+use std::fmt::{Debug, Display};
+use std::future::Future;
+use std::pin::Pin;
 use std::sync::Arc;
+use std::task::{Context, Poll};
 
-use celestia_proto::p2p::pb::{HeaderRequest, HeaderResponse};
+use celestia_proto::p2p::pb::{header_request, HeaderRequest, HeaderResponse};
+use celestia_types::Hash;
+use futures::stream::FuturesUnordered;
+use futures::{FutureExt, Stream};
 use libp2p::{
     request_response::{InboundFailure, RequestId, ResponseChannel},
     PeerId,
 };
-use tracing::instrument;
+use tendermint::hash::Algorithm;
+use tracing::{instrument, trace};
 
+use crate::exchange::utils::{ExtendedHeaderExt, HeaderRequestExt, HeaderResponseExt};
+use crate::exchange::{ReqRespBehaviour, ResponseType};
 use crate::store::Store;
 
-pub(super) struct ExchangeServerHandler<S>
+type StoreJobType<C> = dyn Future<Output = (C, ResponseType)> + Send;
```

Comment:
could `C` be `Channel`? I like single name generics only when I know what they mean :D

Reply:
Tbh we dont need this, we can just use [`futures::future::BoxFuture`](https://docs.rs/futures/latest/futures/future/type.BoxFuture.html) and it will be still readable.

Btw `C` was fine by me. :D

Reply:
BoxFuture is indeed more readable, changed 👍